skada/_utils.py

```diff
@@ -114,13 +172,61 @@ def check_X_domain(
     return_indices: bool = False,
     # xxx(okachaiev): most likely this needs to be removed as it doesn't fit new API
     return_joint: bool = True,
-    allow_auto_sample_domain: bool = False,
+    allow_auto_sample_domain: bool = True,
 ):
+    """
+    Input validation for domain adaptation (DA) estimator.
+    If we work in single-source and single target mode, return source and target
+    separately to avoid additional scan for 'sample_domain' array.
+
+    Parameters:
+    ----------
+    X : array-like of shape (n_samples, n_features)
+        Input features.
+    sample_domain : array-like of shape (n_samples,)
+        Domain labels for each sample.
+    allow_domains : set of int, optional (default=None)
+        Set of allowed domain labels. If provided, only these domain labels are allowed.
+    allow_source : bool, optional (default=True)
+        Allow the presence of source domains.
+    allow_multi_source : bool, optional (default=True)
+        Allow multiple source domains.
+    allow_target : bool, optional (default=True)
+        Allow the presence of target domains.
+    allow_multi_target : bool, optional (default=True)
+        Allow multiple target domains.
+    return_indices : bool, optional (default=False)
+        If True, return only the source indices.
+    return_joint : bool, optional (default=True)
+        If True, return X and sample_domain without separation.
+    allow_auto_sample_domain : bool, optional (default=True)
+        Allow automatic generation of sample_domain if not provided.
+
+    Returns:
+    ----------
+    If return_indices is True:
+        source_idx : array
+            Boolean array indicating source indices.
+    If return_joint is False:
+        X_source : array
+            Input features for source domains.
+        X_target : array
+            Input features for target domains.
+    If return_joint is True:
+        X : array
+            Combined input features for source and target domains.
+        sample_domain : array
+            Combined domain labels for source and target domains.
+    """
     X = check_array(X, input_name='X')
-    if sample_domain is None and allow_auto_sample_domain:
+    if sample_domain is None and not allow_auto_sample_domain:
+        raise ValueError("Either 'sample_domain' or 'allow_auto_sample_domain' "
+                         "should be set")
+    elif sample_domain is None and allow_auto_sample_domain:
         # default target domain when sample_domain is not given
         # xxx(okachaiev): I guess this should be -inf instead of a number
-        sample_domain = -2*np.ones(X.shape[0], dtype=np.int32)
+        # The idea is that with no labels we always assume target domain (-1)
+        sample_domain = -1*np.ones(X.shape[0], dtype=np.int32)
```

Comment:
We put here -1s but we also have `_DEFAULT_TARGET_DOMAIN_LABEL`, which is 2. Should those be the same? 🤔

Reply:
I think that they don't should be the same. 
Here we assume that all datapoints are coming from the same target_domain and that there is no source_domain. Thus we need ONE domain value and it needs to have a negative sign. And if we want to follow the convention of assigning domain values first from (1 to n) and then modifying their sign if its a target domain, we should keep here the -1.

Reply:
Good, I'm okay with this. Let's create a named constant for this situation as well.